File: src/utils/leaderboardCalculations.ts
import type { GameStats } from '../interfaces'
import { REGULAR_SEASON_GAME_COUNT } from '../data/nbaData'
import { LEADERBOARDS } from '../data/nbaLeaderboards'
import type { LeaderboardDef, LeaderboardEntry } from '../data/nbaLeaderboards'

// A season must clear the league rate qualifier (~70% of an 82-game season)
// before its per-game averages can be ranked against NBA seasons.
export const SEASON_AVG_MIN_GAMES = Math.ceil(REGULAR_SEASON_GAME_COUNT * 0.7)

type StatKey = 'points' | 'assists' | 'rebounds' | 'blocks' | 'steals'

const SINGLE_GAME_BOARDS: Partial<Record<string, StatKey>> = {
  'sg-points': 'points',
  'sg-assists': 'assists',
  'sg-rebounds': 'rebounds',
  'sg-blocks': 'blocks',
  'sg-steals': 'steals',
}

const SEASON_AVG_BOARDS: Partial<Record<string, StatKey>> = {
  'season-ppg': 'points',
  'season-rpg': 'rebounds',
  'season-apg': 'assists',
  'season-bpg': 'blocks',
  'season-spg': 'steals',
}

const STAT_ADJECTIVES: Record<StatKey, string> = {
  points: 'point',
  assists: 'assist',
  rebounds: 'rebound',
  blocks: 'block',
  steals: 'steal',
}

export interface LeaderboardRow {
  rank: number
  name: string
  value: number
  context: string
  isPlayer: boolean
}

/** One performance of the tracked player (a game, a season, a career total). */
export interface PlayerPerformance {
  value: number
  context: string
  sortKey: string
  /** 1-based position among the real all-time entries (ties broken by date). */
  rank: number
}

/** The player's best relevant mark, for "your best / not there yet" copy. */
export interface PlayerMarkInfo {
  value: number
  context: string
  sortKey: string
  games: number
  /** Whether the mark is eligible to be ranked (qualified season / complete season). */
  eligible: boolean
  /** All-time rank vs the real board, when meaningful. */
  rank: number | null
}

export interface LeaderboardView {
  board: LeaderboardDef
  rows: LeaderboardRow[]
  /** Player performances merged into the displayed board. */
  playerPerformances: PlayerPerformance[]
  /** Player performances that also make the board but were trimmed for tidiness. */
  extraPlayerCount: number
  mark: PlayerMarkInfo | null
}

export interface TopTwentyEntrance {
  boardId: string
  boardTitle: string
  /** Singular stat word, e.g. "point" or "rebound", for "a 63-point game". */
  adjective: string
  value: number
  context: string
  rank: number
}

const MAX_PLAYER_ROWS_PER_BOARD = 12

const formatGameDate = (date?: string): string => {
  if (!date) {
    return ''
  }
  const parsed = new Date(date)
  if (Number.isNaN(parsed.getTime())) {
    return date
  }
  return parsed.toLocaleDateString(undefined, {
    month: 'short',
    day: 'numeric',
    year: 'numeric',
  })
}

const round2 = (value: number): number => Math.round(value * 100) / 100

/** True when entry `a` outranks `b` (higher value; equal values, earlier sortKey). */
const ranksAbove = (
  a: { value: number; sortKey?: string },
  b: { value: number; sortKey?: string },
): boolean => {
  if (a.value !== b.value) {
    return a.value > b.value
  }
  const aKey = a.sortKey || ''
  const bKey = b.sortKey || ''
  return aKey !== bKey ? aKey < bKey : false
}

const sortedEntries = (board: LeaderboardDef): LeaderboardEntry[] =>
  [...board.entries].sort((a, b) =>
    ranksAbove(a, b) ? -1 : ranksAbove(b, a) ? 1 : 0,
  )

/** The value a performance needs to at least reach to make the board. */
export const getBoardCutoffValue = (board: LeaderboardDef): number => {
  const cutoff = sortedEntries(board)[19]
  return cutoff ? cutoff.value : 0
}

/** 1-based rank `value`/`sortKey` holds among the real all-time entries. */
export const rankAgainstEntries = (
  board: LeaderboardDef,
  value: number,
  sortKey: string,
): number =>
  board.entries.filter((entry) => ranksAbove(entry, { value, sortKey })).length +
  1

// ---------------------------------------------------------------------------
// Candidate selectors — every qualifying performance of the tracked player.
// ---------------------------------------------------------------------------

const playedRegularGames = (games: GameStats[]): GameStats[] =>
  games.filter((game) => !game.isAbsent && game.gameType === 'regular')

type Candidate = PlayerMarkInfo

/** Candidates are ordered best-first by (value desc, sortKey asc). */
const sortCandidates = (candidates: Candidate[]): Candidate[] =>
  [...candidates].sort((a, b) =>
    ranksAbove(a, b) ? -1 : ranksAbove(b, a) ? 1 : 0,
  )

const singleGameCandidates = (
  games: GameStats[],
  stat: StatKey,
): Candidate[] => {
  const played = playedRegularGames(games)
  return played
    .filter((game) => game[stat] > 0)
    .map((game) => ({
      value: game[stat],
      context: `${formatGameDate(game.date)} · ${game.team} vs ${game.opponent}`,
      sortKey: game.date || '',
      games: 1,
      eligible: true,
      rank: 0,
    }))
}

const seasonAverageCandidates = (
  games: GameStats[],
  stat: StatKey,
): Candidate[] => {
  const buckets = new Map<string, GameStats[]>()
  playedRegularGames(games).forEach((game) => {
    const key = `${game.season || ''}::${game.team || ''}`
    const list = buckets.get(key) || []
    list.push(game)
    buckets.set(key, list)
  })

  const candidates: Candidate[] = []
  for (const [key, bucket] of buckets) {
    const season = key.split('::')[0]
    const total = bucket.reduce((sum, game) => sum + game[stat], 0)
    candidates.push({
      value: round2(total / bucket.length),
      context: `${season} season`,
      sortKey: season,
      games: bucket.length,
      eligible: bucket.length >= SEASON_AVG_MIN_GAMES,
      rank: 0,
    })
  }
  return candidates
}

const careerTripleDoubleCandidate = (games: GameStats[]): Candidate | null => {
  const played = playedRegularGames(games)
  const value = played.filter((game) => game.isTripleDouble).length
  if (played.length === 0) {
    return null
  }
  return {
    value,
    context: 'career',
    sortKey: 'career',
    games: played.length,
    eligible: true,
    rank: 0,
  }
}

const seasonTeamWinsCandidates = (games: GameStats[]): Candidate[] => {
  const buckets = new Map<string, GameStats[]>()
  games
    .filter((game) => game.gameType === 'regular')
    .forEach((game) => {
      const key = `${game.season || ''}::${game.team || ''}`
      const list = buckets.get(key) || []
      list.push(game)
      buckets.set(key, list)
    })

  const candidates: Candidate[] = []
  for (const [key, bucket] of buckets) {
    const season = key.split('::')[0]
    candidates.push({
      value: bucket.filter((game) => game.won).length,
      context: `${season} season`,
      sortKey: season,
      games: bucket.length,
      eligible: bucket.length >= REGULAR_SEASON_GAME_COUNT,
      rank: 0,
    })
  }
  return candidates
}

/** All candidate performances for a board, with ranks assigned. */
const boardCandidates = (
  board: LeaderboardDef,
  games: GameStats[],
): Candidate[] => {
  let candidates: Candidate[] = []
  const singleStat = SINGLE_GAME_BOARDS[board.id]
  if (singleStat) {
    candidates = singleGameCandidates(games, singleStat)
  } else if (SEASON_AVG_BOARDS[board.id]) {
    candidates = seasonAverageCandidates(games, SEASON_AVG_BOARDS[board.id] as StatKey)
  } else if (board.id === 'career-triple-doubles') {
    const candidate = careerTripleDoubleCandidate(games)
    candidates = candidate ? [candidate] : []
  } else if (board.id === 'team-season-wins') {
    candidates = seasonTeamWinsCandidates(games)
  }

  return sortCandidates(candidates).map((candidate) => ({
    ...candidate,
    rank: rankAgainstEntries(board, candidate.value, candidate.sortKey),
  }))
}

// ---------------------------------------------------------------------------
// View building — every player performance that cracks a top 20 is merged in.
// ---------------------------------------------------------------------------

const buildView = (
  board: LeaderboardDef,
  stats: GameStats[],
  playerName: string,
): LeaderboardView => {
  const cutoff = getBoardCutoffValue(board)
  const candidates = boardCandidates(board, stats)
  const onBoard = candidates.filter(
    (candidate) => candidate.eligible && candidate.value >= cutoff,
  )
  const extraPlayerCount = Math.max(0, onBoard.length - MAX_PLAYER_ROWS_PER_BOARD)
  const shownPerformances = onBoard.slice(0, MAX_PLAYER_ROWS_PER_BOARD)
  // Best mark worth reporting: the best eligible mark, or the best raw mark
  // when nothing is eligible yet (e.g. an in-progress season).
  const eligibleCandidates = candidates.filter((candidate) => candidate.eligible)
  const mark =
    (eligibleCandidates.length > 0 ? eligibleCandidates : candidates)[0] ||
    null

  const playerRows: { name: string; value: number; context: string; sortKey?: string; isPlayer?: boolean }[] =
    shownPerformances.map((performance) => ({
      name: playerName,
      value: performance.value,
      context: performance.context,
      sortKey: performance.sortKey,
      isPlayer: true,
    }))

  const allRows: {
    name: string
    value: number
    context: string
    sortKey?: string
    isPlayer?: boolean
  }[] = [...board.entries, ...playerRows]

  const rows: LeaderboardRow[] = allRows
    .sort((a, b) => (ranksAbove(a, b) ? -1 : ranksAbove(b, a) ? 1 : 0))
    .map((entry, index) => ({
      rank: index + 1,
      name: entry.name,
      value: entry.value,
      context: entry.context,
      isPlayer: Boolean(entry.isPlayer),
    }))

  return {
    board,
    rows,
    playerPerformances: shownPerformances.map((performance) => ({
      value: performance.value,
      context: performance.context,
      sortKey: performance.sortKey,
      rank: performance.rank ?? 0,
    })),
    extraPlayerCount,
    mark,
  }
}

export const buildLeaderboardViews = (
  stats: GameStats[],
  playerName: string,
): LeaderboardView[] =>
  LEADERBOARDS.map((board) => buildView(board, stats, playerName))

export const formatBoardValue = (
  value: number,
  board: LeaderboardDef,
): string => (board.format === 'dec' ? value.toFixed(2) : String(value))

// ---------------------------------------------------------------------------
// Congrats detection — call right after games are added to the tracker.
// ---------------------------------------------------------------------------

export const computeTopTwentyEntrances = (
  games: GameStats[],
): TopTwentyEntrance[] => {
  const entrances: TopTwentyEntrance[] = []

  games.forEach((game) => {
    if (game.isAbsent || game.gameType !== 'regular') {
      return
    }
    LEADERBOARDS.forEach((board) => {
      const stat = SINGLE_GAME_BOARDS[board.id]
      if (!stat) {
        return
      }
      const value = game[stat]
      if (value <= 0 || value < getBoardCutoffValue(board)) {
        return
      }
      entrances.push({
        boardId: board.id,
        boardTitle: board.title,
        adjective: STAT_ADJECTIVES[stat],
        value,
        context: `${formatGameDate(game.date)} · vs ${game.opponent}`,
        rank: rankAgainstEntries(board, value, game.date || ''),
      })
    })
  })

  return entrances.sort((a, b) => b.value - a.value)
}

File: nba_fantasy_stats_react_app/lib/utils/leaderboard_calculations.dart
import 'package:nba_fantasy_stats_react_app/data/nba_data.dart';
import 'package:nba_fantasy_stats_react_app/data/nba_leaderboards.dart';
import 'package:nba_fantasy_stats_react_app/models/enums.dart';
import 'package:nba_fantasy_stats_react_app/models/game_stats.dart';

/// Dart port of `src/utils/leaderboardCalculations.ts` — ranks the tracked
/// player's performances against real NBA top-20 leaderboards.

/// League rate qualifier for season-average boards — a season must clear
/// ~70% of an 82-game season before its per-game averages can be ranked
/// against NBA seasons. Mirrors `SEASON_AVG_MIN_GAMES` in
/// leaderboardCalculations.ts.
int get seasonAvgMinGames => (regularSeasonGameCount * 0.7).ceil();

/// One of `'points' | 'assists' | 'rebounds' | 'blocks' | 'steals'`.
typedef StatKey = String;

const Map<String, StatKey> _singleGameBoards = {
  'sg-points': 'points',
  'sg-assists': 'assists',
  'sg-rebounds': 'rebounds',
  'sg-blocks': 'blocks',
  'sg-steals': 'steals',
};

const Map<String, StatKey> _seasonAvgBoards = {
  'season-ppg': 'points',
  'season-rpg': 'rebounds',
  'season-apg': 'assists',
  'season-bpg': 'blocks',
  'season-spg': 'steals',
};

const Map<StatKey, String> _statAdjectives = {
  'points': 'point',
  'assists': 'assist',
  'rebounds': 'rebound',
  'blocks': 'block',
  'steals': 'steal',
};

int _statValue(GameStats game, StatKey stat) => switch (stat) {
  'points' => game.points,
  'assists' => game.assists,
  'rebounds' => game.rebounds,
  'blocks' => game.blocks,
  'steals' => game.steals,
  _ => throw ArgumentError('Unknown stat: $stat'),
};

class LeaderboardRow {
  const LeaderboardRow({
    required this.rank,
    required this.name,
    required this.value,
    required this.context,
    required this.isPlayer,
  });

  final int rank;
  final String name;
  final num value;
  final String context;
  final bool isPlayer;
}

/// One performance of the tracked player (a game, a season, a career total).
class PlayerPerformance {
  const PlayerPerformance({
    required this.value,
    required this.context,
    required this.sortKey,
    required this.rank,
  });

  final num value;
  final String context;
  final String sortKey;

  /// 1-based position among the real all-time entries (ties broken by date).
  final int rank;
}

/// The player's best relevant mark, for "your best / not there yet" copy.
class PlayerMarkInfo {
  const PlayerMarkInfo({
    required this.value,
    required this.context,
    required this.sortKey,
    required this.games,
    required this.eligible,
    required this.rank,
  });

  final num value;
  final String context;
  final String sortKey;
  final int games;

  /// Whether the mark is eligible to be ranked (qualified season / complete
  /// season).
  final bool eligible;

  /// All-time rank vs the real board, when meaningful.
  final int? rank;
}

class LeaderboardView {
  const LeaderboardView({
    required this.board,
    required this.rows,
    required this.playerPerformances,
    required this.extraPlayerCount,
    required this.mark,
  });

  final LeaderboardDef board;
  final List<LeaderboardRow> rows;

  /// Player performances merged into the displayed board.
  final List<PlayerPerformance> playerPerformances;

  /// Player performances that also make the board but were trimmed for
  /// tidiness.
  final int extraPlayerCount;
  final PlayerMarkInfo? mark;
}

class TopTwentyEntrance {
  const TopTwentyEntrance({
    required this.boardId,
    required this.boardTitle,
    required this.adjective,
    required this.value,
    required this.context,
    required this.rank,
  });

  final String boardId;
  final String boardTitle;

  /// Singular stat word, e.g. "point" or "rebound", for "a 63-point game".
  final String adjective;
  final num value;
  final String context;
  final int rank;
}

const int _maxPlayerRowsPerBoard = 12;

String _formatGameDate(String? date) {
  if (date == null || date.isEmpty) return '';
  final parsed = DateTime.tryParse(date);
  if (parsed == null) return date;
  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
  ];
  return '${months[parsed.month - 1]} ${parsed.day}, ${parsed.year}';
}

num _round2(num value) => (value * 100).round() / 100;

// ---------------------------------------------------------------------------
// Ranking helpers
// ---------------------------------------------------------------------------

/// Minimal ranking key shared by real entries, player candidates, and merged
/// rows (mirrors the `{ value, sortKey? }` structural type in the TS source).
typedef _RankedKey = ({num value, String sortKey});

/// True when entry [a] outranks [b] (higher value; equal values, earlier
/// sortKey).
bool _ranksAbove(_RankedKey a, _RankedKey b) {
  if (a.value != b.value) {
    return a.value > b.value;
  }
  return a.sortKey != b.sortKey ? a.sortKey.compareTo(b.sortKey) < 0 : false;
}

/// Stable best-first sort (value desc, sortKey asc, full ties keep insertion
/// order) — replicates the stable `Array.prototype.sort` semantics used by
/// the TS source.
List<T> _rankSort<T>(Iterable<T> items, _RankedKey Function(T) keyOf) {
  final indexed = [for (final (index, item) in items.indexed) (index, item)];
  indexed.sort((a, b) {
    if (_ranksAbove(keyOf(a.$2), keyOf(b.$2))) return -1;
    if (_ranksAbove(keyOf(b.$2), keyOf(a.$2))) return 1;
    return a.$1.compareTo(b.$1);
  });
  return [for (final entry in indexed) entry.$2];
}

_RankedKey _entryKey(LeaderboardEntry entry) =>
    (value: entry.value, sortKey: entry.sortKey);

List<LeaderboardEntry> _sortedEntries(LeaderboardDef board) =>
    _rankSort(board.entries, _entryKey);

/// The value a performance needs to at least reach to make the board.
num getBoardCutoffValue(LeaderboardDef board) =>
    _sortedEntries(board).elementAtOrNull(19)?.value ?? 0;

/// 1-based rank `value`/`sortKey` holds among the real all-time entries.
int rankAgainstEntries(LeaderboardDef board, num value, String sortKey) =>
    board.entries
        .where(
          (entry) =>
              _ranksAbove(_entryKey(entry), (value: value, sortKey: sortKey)),
        )
        .length +
    1;

// ---------------------------------------------------------------------------
// Candidate selectors — every qualifying performance of the tracked player.
// ---------------------------------------------------------------------------

typedef _Candidate = ({
  num value,
  String context,
  String sortKey,
  int games,
  bool eligible,
});

List<GameStats> _playedRegularGames(List<GameStats> games) => games
    .where((game) => !game.isAbsent && game.gameType == GameType.regular)
    .toList();

List<_Candidate> _singleGameCandidates(
  List<GameStats> games,
  StatKey stat,
) => _playedRegularGames(games)
    .where((game) => _statValue(game, stat) > 0)
    .map(
      (game) => (
        value: _statValue(game, stat),
        context:
            '${_formatGameDate(game.date)} · ${game.team} vs ${game.opponent}',
        sortKey: game.date.isNotEmpty ? game.date : '',
        games: 1,
        eligible: true,
      ),
    )
    .toList();

List<_Candidate> _seasonAverageCandidates(List<GameStats> games, StatKey stat) {
  final buckets = <String, List<GameStats>>{};
  for (final game in _playedRegularGames(games)) {
    final key = '${game.season}::${game.team}';
    buckets.putIfAbsent(key, () => []).add(game);
  }

  final candidates = <_Candidate>[];
  for (final bucket in buckets.values) {
    final season = bucket.first.season;
    final total = bucket.fold<int>(
      0,
      (sum, game) => sum + _statValue(game, stat),
    );
    candidates.add((
      value: _round2(total / bucket.length),
      context: '$season season',
      sortKey: season,
      games: bucket.length,
      eligible: bucket.length >= seasonAvgMinGames,
    ));
  }
  return candidates;
}

_Candidate? _careerTripleDoubleCandidate(List<GameStats> games) {
  final played = _playedRegularGames(games);
  if (played.isEmpty) return null;
  final value = played.where((game) => game.isTripleDouble).length;
  return (
    value: value,
    context: 'career',
    sortKey: 'career',
    games: played.length,
    eligible: true,
  );
}

List<_Candidate> _seasonTeamWinsCandidates(List<GameStats> games) {
  final buckets = <String, List<GameStats>>{};
  for (final game in games.where((game) => game.gameType == GameType.regular)) {
    final key = '${game.season}::${game.team}';
    buckets.putIfAbsent(key, () => []).add(game);
  }

  final candidates = <_Candidate>[];
  for (final bucket in buckets.values) {
    final season = bucket.first.season;
    candidates.add((
      value: bucket.where((game) => game.won).length,
      context: '$season season',
      sortKey: season,
      games: bucket.length,
      eligible: bucket.length >= regularSeasonGameCount,
    ));
  }
  return candidates;
}

/// All candidate performances for a board, ordered best-first.
List<_Candidate> _boardCandidates(LeaderboardDef board, List<GameStats> games) {
  var candidates = const <_Candidate>[];
  final singleStat = _singleGameBoards[board.id];
  if (singleStat != null) {
    candidates = _singleGameCandidates(games, singleStat);
  } else if (_seasonAvgBoards.containsKey(board.id)) {
    candidates = _seasonAverageCandidates(games, _seasonAvgBoards[board.id]!);
  } else if (board.id == 'career-triple-doubles') {
    final candidate = _careerTripleDoubleCandidate(games);
    candidates = candidate == null ? const <_Candidate>[] : [candidate];
  } else if (board.id == 'team-season-wins') {
    candidates = _seasonTeamWinsCandidates(games);
  }

  return _rankSort(
    candidates,
    (candidate) => (value: candidate.value, sortKey: candidate.sortKey),
  );
}

// ---------------------------------------------------------------------------
// View building — every player performance that cracks a top 20 is merged in.
// ---------------------------------------------------------------------------

LeaderboardView _buildView(
  LeaderboardDef board,
  List<GameStats> stats,
  String playerName,
) {
  final cutoff = getBoardCutoffValue(board);
  final candidates = _boardCandidates(board, stats);
  final onBoard = candidates
      .where((candidate) => candidate.eligible && candidate.value >= cutoff)
      .toList();
  final extraPlayerCount = onBoard.length > _maxPlayerRowsPerBoard
      ? onBoard.length - _maxPlayerRowsPerBoard
      : 0;
  final shownPerformances = onBoard.take(_maxPlayerRowsPerBoard).toList();

  // Best mark worth reporting: the best eligible mark, or the best raw mark
  // when nothing is eligible yet (e.g. an in-progress season).
  final eligibleCandidates = candidates
      .where((candidate) => candidate.eligible)
      .toList();
  final bestPool = eligibleCandidates.isNotEmpty
      ? eligibleCandidates
      : candidates;
  final best = bestPool.isEmpty ? null : bestPool.first;
  final mark = best == null
      ? null
      : PlayerMarkInfo(
          value: best.value,
          context: best.context,
          sortKey: best.sortKey,
          games: best.games,
          eligible: best.eligible,
          rank: best.eligible
              ? rankAgainstEntries(board, best.value, best.sortKey)
              : null,
        );

  final allRows = _rankSort([
    ...board.entries.map(
      (entry) => (
        name: entry.name,
        value: entry.value,
        context: entry.context,
        sortKey: entry.sortKey,
        isPlayer: false,
      ),
    ),
    ...shownPerformances.map(
      (performance) => (
        name: playerName,
        value: performance.value,
        context: performance.context,
        sortKey: performance.sortKey,
        isPlayer: true,
      ),
    ),
  ], (row) => (value: row.value, sortKey: row.sortKey));

  final rows = [
    for (final (index, row) in allRows.indexed)
      LeaderboardRow(
        rank: index + 1,
        name: row.name,
        value: row.value,
        context: row.context,
        isPlayer: row.isPlayer,
      ),
  ];

  return LeaderboardView(
    board: board,
    rows: rows,
    playerPerformances: [
      for (final performance in shownPerformances)
        PlayerPerformance(
          value: performance.value,
          context: performance.context,
          sortKey: performance.sortKey,
          rank: rankAgainstEntries(
            board,
            performance.value,
            performance.sortKey,
          ),
        ),
    ],
    extraPlayerCount: extraPlayerCount,
    mark: mark,
  );
}

List<LeaderboardView> buildLeaderboardViews(
  List<GameStats> stats,
  String playerName,
) => leaderboards.map((board) => _buildView(board, stats, playerName)).toList();

String formatBoardValue(num value, LeaderboardDef board) =>
    board.format == 'dec' ? value.toStringAsFixed(2) : '$value';

// ---------------------------------------------------------------------------
// Congrats detection — call right after games are added to the tracker.
// ---------------------------------------------------------------------------

List<TopTwentyEntrance> computeTopTwentyEntrances(List<GameStats> games) {
  final entrances = <TopTwentyEntrance>[];

  for (final game in games) {
    if (game.isAbsent || game.gameType != GameType.regular) {
      continue;
    }
    for (final board in leaderboards) {
      final stat = _singleGameBoards[board.id];
      if (stat == null) {
        continue;
      }
      final value = _statValue(game, stat);
      if (value <= 0 || value < getBoardCutoffValue(board)) {
        continue;
      }
      entrances.add(
        TopTwentyEntrance(
          boardId: board.id,
          boardTitle: board.title,
          adjective: _statAdjectives[stat] ?? stat,
          value: value,
          context: '${_formatGameDate(game.date)} · vs ${game.opponent}',
          rank: rankAgainstEntries(board, value, game.date),
        ),
      );
    }
  }

  // Sort best-first by value only (stable for ties), matching the TS source.
  final indexed = [for (final (i, e) in entrances.indexed) (i, e)];
  indexed.sort((a, b) {
    final byValue = b.$2.value.compareTo(a.$2.value);
    return byValue != 0 ? byValue : a.$1.compareTo(b.$1);
  });
  return [for (final entry in indexed) entry.$2];
}
